File: mobile/lib/features/ai/games/letter_puzzle/jigsaw_geometry.dart
/// Procedural jigsaw geometry — language-agnostic.
///
/// Instead of hand-drawing each letter's pieces, we lay a rows×cols grid over
/// the letter glyph and generate classic jigsaw edges (tab / blank) whose
/// neighbours always match. This scales to any letter in any script (Arabic,
/// French, English, digits…) with zero per-letter authoring.
///
/// Pure geometry: no widgets, no I/O — unit-testable.
library;

import 'dart:math';
import 'dart:ui';

/// One side of a piece. A `tabOut` on one piece is always paired with a
/// `tabIn` on its neighbour, so adjacent edges interlock.
enum JigsawEdge { flat, tabOut, tabIn }

/// A consistent set of edges for a rows×cols grid. Internal edges are shared:
/// the right edge of (r,c) is the mirror of the left edge of (r,c+1), etc.
class JigsawLayout {
  final int rows;
  final int cols;

  /// Vertical cuts between columns: `_vEdge[r][c]` is the edge on the RIGHT of
  /// piece (r,c) for c in 0..cols-2. `tabOut` means (r,c) bulges right.
  final List<List<JigsawEdge>> _vEdge;

  /// Horizontal cuts between rows: `_hEdge[r][c]` is the edge on the BOTTOM of
  /// piece (r,c) for r in 0..rows-2. `tabOut` means (r,c) bulges down.
  final List<List<JigsawEdge>> _hEdge;

  JigsawLayout._(this.rows, this.cols, this._vEdge, this._hEdge);

  /// Generate a layout with random (but neighbour-consistent) knobs.
  factory JigsawLayout.generate(int rows, int cols, {int? seed}) {
    final rnd = Random(seed);
    final vEdge = List.generate(
      rows,
      (_) => List.generate(
        cols - 1,
        (_) => rnd.nextBool() ? JigsawEdge.tabOut : JigsawEdge.tabIn,
      ),
    );
    final hEdge = List.generate(
      rows - 1,
      (_) => List.generate(
        cols,
        (_) => rnd.nextBool() ? JigsawEdge.tabOut : JigsawEdge.tabIn,
      ),
    );
    return JigsawLayout._(rows, cols, vEdge, hEdge);
  }

  static JigsawEdge _mirror(JigsawEdge e) => switch (e) {
        JigsawEdge.tabOut => JigsawEdge.tabIn,
        JigsawEdge.tabIn => JigsawEdge.tabOut,
        JigsawEdge.flat => JigsawEdge.flat,
      };

  JigsawEdge topOf(int r, int c) =>
      r == 0 ? JigsawEdge.flat : _mirror(_hEdge[r - 1][c]);
  JigsawEdge bottomOf(int r, int c) =>
      r == rows - 1 ? JigsawEdge.flat : _hEdge[r][c];
  JigsawEdge leftOf(int r, int c) =>
      c == 0 ? JigsawEdge.flat : _mirror(_vEdge[r][c - 1]);
  JigsawEdge rightOf(int r, int c) =>
      c == cols - 1 ? JigsawEdge.flat : _vEdge[r][c];

  int get pieceCount => rows * cols;
}

/// How far a knob bulges, as a fraction of the shorter cell side.
const double _kKnobFraction = 0.22;

/// Neck width of a knob, as a fraction of the edge length.
const double _kNeck = 0.18;

/// Build the closed [Path] for the piece at (row,col).
///
/// Coordinates are in BOARD space: the piece sits at its grid cell, and knobs
/// may extend beyond the cell rectangle into neighbouring cells (that overlap
/// is intentional and how jigsaws interlock).
Path jigsawPiecePath({
  required JigsawLayout layout,
  required int row,
  required int col,
  required double cellW,
  required double cellH,
}) {
  final left = col * cellW;
  final top = row * cellH;
  final right = left + cellW;
  final bottom = top + cellH;
  final knob = _kKnobFraction * min(cellW, cellH);

  final path = Path()..moveTo(left, top);

  _edge(
    path,
    layout.topOf(row, col),
    Offset(left, top),
    Offset(right, top),
    cellW,
    knob,
    outwardSign: -1,
    horizontal: true,
  );
  _edge(
    path,
    layout.rightOf(row, col),
    Offset(right, top),
    Offset(right, bottom),
    cellH,
    knob,
    outwardSign: 1,
    horizontal: false,
  );
  _edge(
    path,
    layout.bottomOf(row, col),
    Offset(right, bottom),
    Offset(left, bottom),
    cellW,
    knob,
    outwardSign: 1,
    horizontal: true,
  );
  _edge(
    path,
    layout.leftOf(row, col),
    Offset(left, bottom),
    Offset(left, top),
    cellH,
    knob,
    outwardSign: -1,
    horizontal: false,
  );

  path.close();
  return path;
}

/// Draw one edge from [start] to [end]. For tab edges, a rounded knob bulges
/// out (away from the piece centre) for [JigsawEdge.tabOut] and in for tabIn.
void _edge(
  Path path,
  JigsawEdge edge,
  Offset start,
  Offset end,
  double length,
  double knob, {
  required int outwardSign,
  required bool horizontal,
}) {
  if (edge == JigsawEdge.flat) {
    path.lineTo(end.dx, end.dy);
    return;
  }

  // Direction along the edge and the outward normal.
  final dir = (end - start) / length;
  final normal = horizontal
      ? Offset(0, outwardSign.toDouble())
      : Offset(outwardSign.toDouble(), 0);
  final bulge = edge == JigsawEdge.tabOut ? normal * knob : normal * -knob;

  final neckStart = start + dir * (length * (0.5 - _kNeck));
  final neckEnd = start + dir * (length * (0.5 + _kNeck));
  final headCenter = start + dir * (length * 0.5) + bulge;
  final headLeft = neckStart + bulge;
  final headRight = neckEnd + bulge;

  path.lineTo(neckStart.dx, neckStart.dy);
  // Up into the knob, around the head, and back down — two cubics.
  path.cubicTo(
    headLeft.dx,
    headLeft.dy,
    headCenter.dx,
    headCenter.dy,
    headCenter.dx,
    headCenter.dy,
  );
  path.cubicTo(
    headCenter.dx,
    headCenter.dy,
    headRight.dx,
    headRight.dy,
    neckEnd.dx,
    neckEnd.dy,
  );
  path.lineTo(end.dx, end.dy);
}

/// The straight cell rectangle for a slot at (row,col) — used to position the
/// recessed board slots (the knob overlap is handled by the piece path).
Rect cellRect(int row, int col, double cellW, double cellH) =>
    Rect.fromLTWH(col * cellW, row * cellH, cellW, cellH);

File: web/src/features/ai/games/lib/jigsawGeometry.ts
/**
 * Procedural jigsaw geometry (web) — mirrors the Flutter `jigsaw_geometry.dart`.
 *
 * Lays a rows×cols grid over a glyph and generates classic tab/blank jigsaw
 * edges whose neighbours always interlock. Produces an SVG path string per
 * piece, so it scales to any letter / language with no per-letter authoring.
 *
 * Each piece is returned in LOCAL coordinates with the cell inset by `pad`
 * (= knob size) on every side, so all tiles share the same size and align by
 * positioning the cell, not the bounds.
 */

export type Edge = 'flat' | 'tabOut' | 'tabIn';

const KNOB_FRACTION = 0.22;
const NECK = 0.18;

function mirror(e: Edge): Edge {
  if (e === 'tabOut') return 'tabIn';
  if (e === 'tabIn') return 'tabOut';
  return 'flat';
}

// Tiny deterministic PRNG (mulberry32) so layouts are stable per seed.
function rng(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export interface JigsawLayout {
  rows: number;
  cols: number;
  top: (r: number, c: number) => Edge;
  right: (r: number, c: number) => Edge;
  bottom: (r: number, c: number) => Edge;
  left: (r: number, c: number) => Edge;
}

export function generateLayout(rows: number, cols: number, seed: number): JigsawLayout {
  const rnd = rng(seed);
  // vertical cuts: vEdge[r][c] = right edge of (r,c), c in 0..cols-2
  const vEdge: Edge[][] = Array.from({ length: rows }, () =>
    Array.from({ length: Math.max(0, cols - 1) }, () => (rnd() < 0.5 ? 'tabOut' : 'tabIn')),
  );
  // horizontal cuts: hEdge[r][c] = bottom edge of (r,c), r in 0..rows-2
  const hEdge: Edge[][] = Array.from({ length: Math.max(0, rows - 1) }, () =>
    Array.from({ length: cols }, () => (rnd() < 0.5 ? 'tabOut' : 'tabIn')),
  );
  return {
    rows,
    cols,
    top: (r, c) => (r === 0 ? 'flat' : mirror(hEdge[r - 1][c])),
    bottom: (r, c) => (r === rows - 1 ? 'flat' : hEdge[r][c]),
    left: (r, c) => (c === 0 ? 'flat' : mirror(vEdge[r][c - 1])),
    right: (r, c) => (c === cols - 1 ? 'flat' : vEdge[r][c]),
  };
}

interface Pt {
  x: number;
  y: number;
}

function edgeTo(
  parts: string[],
  edge: Edge,
  start: Pt,
  end: Pt,
  length: number,
  knob: number,
  outwardSign: number,
  horizontal: boolean,
): void {
  if (edge === 'flat') {
    parts.push(`L ${end.x} ${end.y}`);
    return;
  }
  const dir: Pt = { x: (end.x - start.x) / length, y: (end.y - start.y) / length };
  const normal: Pt = horizontal ? { x: 0, y: outwardSign } : { x: outwardSign, y: 0 };
  const sign = edge === 'tabOut' ? 1 : -1;
  const bulge: Pt = { x: normal.x * knob * sign, y: normal.y * knob * sign };

  const at = (t: number): Pt => ({ x: start.x + dir.x * length * t, y: start.y + dir.y * length * t });
  const neckStart = at(0.5 - NECK);
  const neckEnd = at(0.5 + NECK);
  const mid = at(0.5);
  const head: Pt = { x: mid.x + bulge.x, y: mid.y + bulge.y };
  const headLeft: Pt = { x: neckStart.x + bulge.x, y: neckStart.y + bulge.y };
  const headRight: Pt = { x: neckEnd.x + bulge.x, y: neckEnd.y + bulge.y };

  parts.push(`L ${neckStart.x} ${neckStart.y}`);
  parts.push(`C ${headLeft.x} ${headLeft.y} ${head.x} ${head.y} ${head.x} ${head.y}`);
  parts.push(`C ${head.x} ${head.y} ${headRight.x} ${headRight.y} ${neckEnd.x} ${neckEnd.y}`);
  parts.push(`L ${end.x} ${end.y}`);
}

export interface PieceGeom {
  row: number;
  col: number;
  d: string; // SVG path in local coords (cell inset by pad)
  tileW: number;
  tileH: number;
  pad: number;
  cellW: number;
  cellH: number;
}

/** Build the local SVG path for piece (row,col). */
export function piecePath(
  layout: JigsawLayout,
  row: number,
  col: number,
  cellW: number,
  cellH: number,
): PieceGeom {
  const knob = KNOB_FRACTION * Math.min(cellW, cellH);
  const pad = knob;
  const l = pad;
  const t = pad;
  const r = pad + cellW;
  const b = pad + cellH;

  const parts: string[] = [`M ${l} ${t}`];
  edgeTo(parts, layout.top(row, col), { x: l, y: t }, { x: r, y: t }, cellW, knob, -1, true);
  edgeTo(parts, layout.right(row, col), { x: r, y: t }, { x: r, y: b }, cellH, knob, 1, false);
  edgeTo(parts, layout.bottom(row, col), { x: r, y: b }, { x: l, y: b }, cellW, knob, 1, true);
  edgeTo(parts, layout.left(row, col), { x: l, y: b }, { x: l, y: t }, cellH, knob, -1, false);
  parts.push('Z');

  return {
    row,
    col,
    d: parts.join(' '),
    tileW: cellW + 2 * pad,
    tileH: cellH + 2 * pad,
    pad,
    cellW,
    cellH,
  };
}

export function allPieces(
  layout: JigsawLayout,
  cellW: number,
  cellH: number,
): PieceGeom[] {
  const out: PieceGeom[] = [];
  for (let r = 0; r < layout.rows; r++) {
    for (let c = 0; c < layout.cols; c++) {
      out.push(piecePath(layout, r, c, cellW, cellH));
    }
  }
  return out;
}
